src/expr/src/linear.rs

```diff
@@ -889,23 +889,40 @@ impl MapFilterProject {
     /// );
     /// ```
     pub fn optimize(&mut self) {
-        // Optimization memoizes individual `ScalarExpr` expressions that
-        // are sure to be evaluated, canonicalizes references to the first
-        // occurrence of each, inlines expressions that have a reference
-        // count of one, and then removes any expressions that are not
-        // referenced.
-        self.memoize_expressions();
-        self.predicates.sort();
-        self.predicates.dedup();
-        self.inline_expressions();
-        self.remove_undemanded();
-
-        // Re-build `self` from parts to restore evaluation order invariants.
-        let (map, filter, project) = self.as_map_filter_project();
-        *self = Self::new(self.input_arity)
-            .map(map)
-            .filter(filter)
-            .project(project);
+        // Track sizes and iterate as long as they decrease.
+        let mut prev_size = usize::max_value();
+        let mut self_size = self.size();
```

Comment:
Can probably reframe this to avoid the initial size calculation, because we never intend to read it but just to start iteration unconditionally.